Add SOME/IP SD protocol

kolbex opened the pull request:
Add SOME/IP SD layer support
    - Add SomeIpSdLayer class for handling SOME/IP SD packets
    - Options: Endpoint/Multicast for IPv4/IPv6, Configuration option
    - Entry types: Service/Eventgroup Entry Type
    - Add tests

seladb commented:
@kolbex pre-commit fails because of spelling issues, can you please fix them?

kolbex commented:
@seladb can you please do a first review on SOME/IP-SD. It is currently not implemented with TLV. SOME/IP-SD provides entries which has a type field and data (no length filed) and Options which has a length field + type field and data. Do you suggest to implent TLV because it does not fit exactly to the TLVRecord<>?

seladb commented:
@kolbex can you please resolve a few issues before reviewing your PR:

- The PR conflicts with `dev`. The reason is https://github.com/seladb/PcapPlusPlus/pull/936 was merged which contains a new protocol (WoL)
- please notice CI is failing because of  the following doxygen documentation error:
   ```text
   Generating docs for compound pcpp::RapidStpLayer.../tmp/cirrus-ci-build/Dist/header/SomeIpLayer.h:90: error: unable to    resolve reference to `someiphdr' for \ref command (warning treated as error, aborting now)
   ```
   https://github.com/seladb/PcapPlusPlus/pull/911/checks?check_run_id=8677258968

seladb commented on Packet++/header/SomeIpSdLayer.h:
```diff
@@ -0,0 +1,998 @@
+	 * @param[in] minorVersion MinorVersion to use
+	 * @return std::unique_ptr<SomeIpSdEntry>
+	 */
+	static std::unique_ptr<SomeIpSdEntry> create(EntryType type, uint16_t serviceID, uint16_t instanceID,
+												 uint8_t majorVersion, uint32_t TTL, uint32_t minorVersion);
+
+	/**
+	 * Creates an entry of the right type and returns it as a base pointer. Used to create SomeIpSd
+	 * EntryType 2 objects.
+	 * @param[in] type Type to create
+	 * @param[in] serviceID ServiceID to use
+	 * @param[in] instanceID InstanceID to use
+	 * @param[in] majorVersion MajorVersion to use
+	 * @param[in] TTL TTL to use. Has to be 0 for all Stop* entry types
+	 * @param[in] counter Counter value to use
+	 * @param[in] eventGroupID EventgroupId to use
+	 * @return std::unique_ptr<SomeIpSdEntry>
+	 */
+	static std::unique_ptr<SomeIpSdEntry> create(EntryType type, uint16_t serviceID, uint16_t instanceID,
```
I wonder if we need this static `create(...)` methods. The c'tor is public so the user can just call it to create an object, no?

seladb commented on Packet++/header/SomeIpSdLayer.h:
```diff
@@ -0,0 +1,998 @@
+	 * @return TEntryPtr& Returns a pointer to the entry in the layer. Note that this pointer will be invalid as
+	 * soon as an further option or entry is added to the layer.
+	 */
+	template <class TEntryPtr> TEntryPtr &addEntry(const TEntryPtr &pEntry);
```
I suggest this method doesn't accept a pointer but a const reference (`const SomeIpSdEntry&`) and just copies its data to the packet

kolbex commented on Packet++/header/SomeIpLayer.h:
```diff
@@ -133,7 +133,7 @@ class SomeIpLayer : public Layer
 	/**
 	 * Get a pointer to the basic SOME/IP header. Notice this points directly to the data, so every change will change
 	 * the actual packet data
-	 * @return A pointer to the @ref someiphdr
```
because of the failing test from the Doxygen Documentation pipeline (also line 90)...not sure why...it already worked with the SOME/IP-TP branch...for me the `@ref someiphdr` looks ok:
`Generating docs for compound pcpp::radius_hea/tmp/cirrus-ci-build/Dist/header/SomeIpLayer.h:136: error: unable to resolve reference to `someiphdr' for \ref command (warning treated as error, aborting now)`

seladb replied:
Hmm I'm not sure why you get this error 🤔 
I'll take a look...

kolbex commented on Packet++/header/SomeIpSdLayer.h:
```diff
@@ -0,0 +1,998 @@
+	 * @param[in] port Port to check
+	 * @return true if SomeIP protocol port, false if not
+	 */
+	static bool isSomeIpSdPort(uint16_t port) { return port == 30490; }
```
this was an open issue which i fixed now...SD communication can only be on port 30490...the testcase was wrong

seladb commented:
Hi @kolbex , I'm really sorry for the delay. I see you just added changes and the build is failing. Please let me know when it's fixed and I'll try to re-review the PR asap

kolbex commented:
@seladb i did a small update, one android compiler is still failing, but the others are passing, can you please take a look on this PR?

seladb commented:
@kolbex I'm not sure which changes you made, but the Android build on mac fails with this error:

```shell
/Users/runner/work/PcapPlusPlus/PcapPlusPlus/ToyVpn-PcapPlusPlus/app/src/main/cpp/../../../libs/pcapplusplus/arm64-v8a/30/libPacket++.a(SomeIpSdLayer.o): In function `long std::__ndk1::__libcpp_atomic_refcount_decrement<long>(long&)':
/Users/runner/Library/Android/sdk/ndk/25.1.8937393/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include/c++/v1/memory:3450: undefined reference to `__aarch64_ldadd8_acq_rel'
/Users/runner/Library/Android/sdk/ndk/25.1.8937393/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include/c++/v1/memory:3450: undefined reference to `__aarch64_ldadd8_acq_rel'
/Users/runner/Library/Android/sdk/ndk/25.1.8937393/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include/c++/v1/memory:3450: undefined reference to `__aarch64_ldadd8_acq_rel'
/Users/runner/Library/Android/sdk/ndk/25.1.8937393/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include/c++/v1/memory:3450: undefined reference to `__aarch64_ldadd8_acq_rel'
/Users/runner/Library/Android/sdk/ndk/25.1.8937393/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include/c++/v1/memory:3450: undefined reference to `__aarch64_ldadd8_acq_rel'
/Users/runner/work/PcapPlusPlus/PcapPlusPlus/ToyVpn-PcapPlusPlus/app/src/main/cpp/../../../libs/pcapplusplus/arm64-v8a/30/libPacket++.a(SomeIpSdLayer.o):/Users/runner/Library/Android/sdk/ndk/25.1.8937393/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include/c++/v1/memory:3450: more undefined references to `__aarch64_ldadd8_acq_rel' follow
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

Any idea which change might cause this? 🤔

seladb commented:
@kolbex I searched a little bit and maybe it has to do with the `std::unique_ptr` you're using?

Here are a few threads that might be related:
https://www.mail-archive.com/debian-bugs-dist@lists.debian.org/msg1802783.html
https://forum.yoyogames.com/index.php?threads/unable-to-yyc-compile-android-application.98979/
https://github.com/termux/termux-packages/issues/4760

Maybe we can solve it by adding `-latomic` or ` -mno-outline-atomics` build flags (I'm not sure, I haven't tested it), but I don't think it is worth it... if `std::unique_ptr` is the problem, we should probably find a way around it

seladb commented on Packet++/header/SomeIpSdLayer.h:
```diff
@@ -0,0 +1,771 @@
+	 * Represents the Load Balancing option header for SOME/IP-SD
+	 */
+#pragma pack(push, 1)
+	struct someipsdhdroptionsload
```
Here also: `someipsdhdroptionsload` can inherit from `someipsdhdroptionsbase`

seladb commented on Packet++/src/SomeIpSdLayer.cpp:
```diff
@@ -0,0 +1,789 @@
+		return true;
+	}
+
+	uint32_t indexSecondOption = (uint32_t)hdrEntry->indexSecondOption;
+	uint32_t lenSecondOption = (uint32_t)hdrEntry->nrOpt2;
```
why casting to `uint32_t`? I think it'll be easier to stay with `uint8_t`, no?

kolbex replied:
yes, not needed any more.

kolbex commented on Packet++/header/SomeIpSdLayer.h:
```diff
@@ -0,0 +1,771 @@
+	/**
+	 * Types of options currently available for the SOME/IP-SD protocol
+	 */
+	enum class OptionType : uint8_t
```
i added the 'Unknown' option type, but i think this is not needed currently

seladb replied:
It might be used here: https://github.com/kolbex/PcapPlusPlus/blob/ac36a0c9bc953b04ad8ba72f378503a8572cd491/Packet%2B%2B/src/SomeIpSdLayer.cpp#L22

If `getSomeIpSdOptionHeader()->type` has an invalid value

kolbex commented:
@seladb sorry for the late reply, i think you are right and the shared_ptr is the cause for the problem with the android build error. In the previous version i used unique_ptr but there were compiler errors so i changed to shared_ptr. Sorry for rebasing which makes it harder to follow the changes. Anyway it was a very huge change to remove the m_Entries and m_Options members. If you want the version before i can send it to you? Currently i removed the shared_ptr from the EntryPtr and OptionPtr to become error free but i hope we can find a better solution for cleaning up the vector from getEntries() and getOptions(). Do you have a good idea?
